subsurface-mobile-main.cpp

```diff
@@ -19,14 +19,48 @@
 #include <QLocale>
 #include <git2.h>
 
+// Implementation of STP logging 
+#include "core/ssrf.h"
+#ifdef ENABLE_STARTUP_TIMING
+#include <QElapsedTimer>
+void log_stp(const char *ident, void *buf)
```

Comment:
well Qstring cannot be used, because it would demand that all cpp sources include QString, which is not guaranteed.

Same goes for the void *.

btw this is a frequently used methid to generalize extern functions and avoid making demands on the including source.

Reply:
> well Qstring cannot be used, because it would demand that all cpp sources include QString, which is not guaranteed.

The function declaration is in a `#ifdef __cplusplus` block.
I don't see why you couldn't simply put a
```
class QString;
```
forward-declaration in this block. You could even `#include <QString>` in this block.

Perhaps you have to move it out of the `extern "C"` thing. But I don't see the point of declaring a function as `extern "C"`, when it can only be called from C++. And since you return the log as a `QString`, it can only be called from C++. Which again proves that this should be two functions.

Actually, I think it could even be three:
1) Start timer
2) Log
3) Access log

> btw this is a frequently used methid to generalize extern functions and avoid making demands on the including source.

I think opaque pointers were mostly used in days of yore by libraries to avoid applications accessing private data. To prevent disasters in the case that the data format changed. Qt actually does something similar with its PIMPL idiom. But I see no reason why we would do this internally.

Reply:
ageism... some of us resemble this remark

Reply:
I don't think there's a point in over-designing this code, but I think there are some good points here.
1) the multi entry to the same function confused me as well and I don't see a strong value in it... the downside of that is the need for a global QMutex, but again, this is just some debugging code, not the cradle of the future of Subsurface
2) returning QString from the function that gets us the string just seems much more reasonable; in that sense the suggestion of 
```
void log_stp(const char *);
QString get_stp_log();
```
seems like a good way to go. Or 
```
void logStp(const char *);
QString getStpLog();
```
Again, this is debug code - naming is, err, not my key concern.
3) I really don't care how the string is assembled - .append is fine by me

Reply:
well 3 functions would mean the variables need to be global, something I was told earlier to avoid. I do not understand the confusion, there are 2 macros calling a common function, a pattern you see often (multiple interface functions calling one implementation function).

making a forward declaration to simulate a #include is at least as ugly as a void *

it is not in the extern “C” thing, it is in the cplusplus part to avoid including it in C code, somethung which wax requested in the earlier reviews.

Design can be discussed for a very long time, I used the best design considering the constraint I was given, clearly if the constraints changes others designs are simpler, but hardly fair, and in this case overengineering do not forget this is the second rework.

Should the 3 fubctions then also be inline so we have extra calls in the release code ? (another constraint change)

Reply:
Note that I'm only giving suggestions on how I think the code can be improved. I'm not posing constraints. Even without going over-the-top, I think this can be made much more sensible.

> well 3 functions would mean the variables need to be global

No, they're *local* to the translation unit (of static linkage in C++ parlance).

> making a forward declaration to simulate a #include is at least as ugly as a void *

Sorry, here I strongly disagree. Forward-declarations of types to enable strong typing is orders-of-magnitude less ugly than duck-typing via `void *`. You will find forward-declaration is a very common and often recommended idiom.
See for example: Point #7 in http://www.acodersjourney.com/2016/05/top-10-c-header-file-mistakes-and-how-to-fix-them/

> it is not in the extern “C” thing, it is in the cplusplus part to avoid including it in C code

Well, then all the better.

> Should the 3 fubctions then also be inline so we have extra calls in the release code 

I was trying to suggest something different:
```
#ifdef DEBUG
class QString;
extern void log_internal(...); // Note: extern not needed, since default linkage for functions.
extern QString get_log_internal();
inline void log(...)
{
    log_internal(...);
}
inline QString get_log()
{
    return get_log_internal();
}
#else
inline void log(...)
{
}
inline QString get_log()
{
    return QString();
}
#endif
```

Of course, macros work just as well with the well known caveats.